kyo-scheduler/jvm/src/main/scala/kyo/scheduler/Scheduler.scala

```diff
@@ -0,0 +1,230 @@
+            var i       = XSRandom.nextInt(m)
+            var tries   = Math.min(m, scheduleTries)
+            var minLoad = Int.MaxValue
+            while tries > 0 && minLoad != 0 do
+                val w = workers(i)
+                if w != null && !w.handleBlocking() then
+                    val l = w.load()
+                    if l < minLoad && (w ne submitter) then
+                        minLoad = l
+                        worker = w
+                end if
+                i += 1
+                if i == m then
+                    i = 0
+                tries -= 1
+            end while
+        end if
+        while worker == null do
+            worker = workers(XSRandom.nextInt(maxConcurrency))
+        if !worker.enqueue(t) then
+            schedule(t, submitter)
+    end schedule
+
+    private def steal(thief: Worker): Task =
+        var worker: Worker = null
+        var i              = 0
+        var maxLoad        = Int.MaxValue
+        while i < maxConcurrency do
+            val w = workers(i)
+            if w != null && !w.handleBlocking() then
+                val l = w.load()
+                if l > maxLoad && (w ne thief) then
+                    maxLoad = l
+                    worker = w
+            end if
+            i += 1
+        end while
+        if worker != null then
+            worker.steal(thief)
+        else
+            null
+        end if
+    end steal
+
+    def flush() =
+        val w = Worker.current()
+        if w != null then
+            flushes.increment()
+            w.drain()
+    end flush
+
+    def loadAvg(): Double =
+        val m = this.maxConcurrency
+        var i = 0
+        var r = 0
+        while i < m do
+            val w = workers(i)
+            if w != null then
+                r += w.load()
+            i += 1
+        end while
+        r.toDouble / m
+    end loadAvg
+
+    def shutdown(): Unit =
+        cycleTask.cancel(true)
+        admissionRegulator.stop()
+        concurrencyRegulator.stop()
+        gauges.close()
+    end shutdown
+
+    private def updateWorkers(delta: Int) =
+        maxConcurrency = Math.max(minWorkers, Math.min(maxWorkers, maxConcurrency + delta))
+        ensureWorkers()
+
+    private def ensureWorkers() =
+        for idx <- allocatedWorkers until maxConcurrency do
+            workers(idx) = new Worker(idx, pool, schedule, steal, () => cycles, clock)
+            allocatedWorkers += 1
+
+    private val cycleTask =
+        timerExecutor.scheduleAtFixedRate(
+            () => cycleWorkers(),
+            timeSliceMs,
+            timeSliceMs,
+            TimeUnit.MILLISECONDS
+        )
+
+    private def cycleWorkers(): Unit =
+        try
+            val curr = cycles + 1
+            cycles = curr
+            var i = 0
+            while i < allocatedWorkers do
+                val w = workers(i)
+                if w != null then
+                    if i >= maxConcurrency then
+                        w.drain()
```

Comment:
Improved logic to continue cycling all workers but also draining workers that have been stopped (are above the concurrency limit)